engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,322 @@
+  when(LoadingCache) {
+    case Event(request: DockerHashRequest, data) =>
+      stay using data.addHashRequest(request, sender())
+    case Event(DockerHashStoreLoadingSuccess(dockerHashEntries), data) =>
+      loadCacheAndHandleHashRequests(dockerHashEntries, data)
+  }
+
+  // This is the normal operational mode.
+  when(Running) {
+    // This tag has already been looked up and its hash is in the mappings cache.
+    case Event(request: DockerHashRequest, data) if data.mappings.contains(request.dockerImageID) =>
+      sender ! DockerHashSuccessResponse(data.mappings(request.dockerImageID), request)
+      stay()
+    // A request for the hash for this tag has already been made to the hashing actor.  Don't request the hash again,
+    // just add this sender to the list of replyTos for when the hash arrives.
+    case Event(request: DockerHashRequest, data) if data.hashRequests.contains(request) =>
+      stay using data.addHashRequest(request, sender())
+    // This tag has not (successfully) been looked up before, so look it up now.
+    case Event(request: DockerHashRequest, data) =>
+      requestDockerHash(request, data)
+    case Event(dockerResponse: DockerHashSuccessResponse, data) =>
+      persistDockerHash(dockerResponse, data)
+    case Event(dockerResponse: DockerHashFailureResponse, data) =>
+      handleLookupFailure(dockerResponse, data)
+    case Event(DockerHashStoreSuccess(response), data) =>
+      recordMappingAndRespond(response, data)
+    case Event(DockerHashStoreFailure(request, e), data) =>
+      handleStoreFailure(request, new Exception(s"Failure storing docker hash for ${request.dockerImageID.fullName}", e), data)
+  }
+
+  // In state Terminal we reject all requests with the cause set in the state data.
+  when(Terminal) {
+    case Event(request: DockerHashRequest, data) =>
+      sender() ! WorkflowDockerLookupFailure(data.failureCause.orNull, request)
+      stay()
+  }
+
+  private def fail(reason: Throwable): State = {
+    self ! TransitionToFailed(reason)
+    stay()
+  }
+
+  whenUnhandled {
+    case Event(DockerHashActorTimeout(message), data) =>
+      // We have no way of knowing the offending request.  The logic below fails all in-flight requests but soldiers on
+      // in the current state.
+      val reason = new Exception(s"Timeout looking up docker hash: $message")
+      val updatedData = respondToAllRequestsWithLookupFailure(reason, data)
+      stay() using updatedData
+    case Event(ShutDown, _) =>
+      databaseInterface.removeDockerHashStoreEntries(workflowId.toString) onComplete {
+        case Success(_) =>
+          self ! TransitionToShutDown
+        case Failure(e) =>
+          // This isn't great in that it leaves a bunch of mappings in the database, but it's not worth failing the
+          // workflow just because we couldn't clean this up.
+          log.error(e, "Failed to remove docker hash store entries for workflow {}", workflowId)
```

Comment:
Right now, we remove workflows [here](https://github.com/broadinstitute/cromwell/blob/develop/engine/src/main/scala/cromwell/engine/workflow/workflowstore/WorkflowStoreEngineActor.scala#L93).

You *might* argue that `WorkflowStore` and `WorkflowHashes` are separated concerns. To which I *might* counter that actually they aren't really and should be deleted together.